<a href="https://colab.research.google.com/github/rahulsikri/rahul_rl/blob/main/Exp_1_Probability_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv('student-mat.csv')
df.head(3)

school sex  age address famsize Pstatus  Medu  Fedu     Mjob     Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home  teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home    other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home    other  ...   

  famrel freetime  goout  Dalc  Walc health absences G1 G2  G3  
0      4        3      4     1     1      3        6  5  6   6  
1      5        3      3     1     1      3        4  5  5   6  
2      4        3      2     2     3      3       10  7  8  10  

[3 rows x 33 columns]

In [ ]:
len(df)

395

Add a boolean column called grade_A noting if a student achieved 80% or higher as a final score. Original values are on a 0–20 scale so we multiply by 5.

In [ ]:
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0    ...      4        3      4     1     1      3        6   5   6   6  
1    ...      5        3      3     1     1      3        4   5   5   6  
2    ...      4        3      2     2     3      3       10   7   8  10  
3    ...      3        2      2     1     1      5        2  15  14  15  
4    ...      4        3      2     1     2      5        4   6  10  10  
..   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
390  ...      5        5      4     4     5      4       11   9   9   9  
391  ...      2        4      5     3     4      2        3  14  16  16  
392  ...      5        5      3     3     3      3        3  10   8   7  
393  ...      4        4      1     3     4      5        0  11  12  10  
394  ...      3        2      3     3     3      5        5   8   9   9  

[395 rows x 33 columns]

In [ ]:
x = df['G3']*5
x

0      30
1      30
2      50
3      75
4      50
       ..
390    45
391    80
392    35
393    50
394    45
Name: G3, Length: 395, dtype: int64

In [ ]:
pivot = df.iloc[0:395][['absences','G3']]
pivot

absences  G3
0           6   6
1           4   6
2          10  10
3           2  15
4           4  10
..        ...  ..
390        11   9
391         3  16
392         3   7
393         0  10
394         5   9

[395 rows x 2 columns]

In [ ]:
count = len(pivot[(pivot['absences']>10) &(pivot['G3']>16)])
print(count)

3


In [ ]:
count1 = len(pivot[pivot['absences']>10])
print(count1)

66


In [ ]:
count2 = len(pivot[(pivot['G3']>16)])
print(count2)

24


In [ ]:
P_a_intersection_b = count/395
P_a = count2/395
P_b = count1/395

In [ ]:
P = P_a_intersection_b/P_b
P

0.04545454545454545

In [ ]:
print(f"The probability of getting at least an 80% final grade, given missing 10 or more classes is {P}")

The probability of getting at least an 80% final grade, given missing 10 or more classes is 0.04545454545454545


## Second Way to implement

In [ ]:
df['grade_A'] = np.where(df['G3']*5 >= 80, 1, 0)

Make another boolean column called high_absenses with a value of 1 if a student missed 10 or more classes.

In [ ]:
df['high_absenses'] = np.where(df['absences'] >= 10, 1, 0)

Add one more column to make building a pivot table easier.

In [ ]:
df['count'] = 1

In [ ]:
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... Dalc Walc  health  absences  G1  G2  G3 grade_A high_absenses count  
0    ...    1    1       3         6   5   6   6       0             0     1  
1    ...    1    1       3         4   5   5   6       0             0     1  
2    ...    2    3       3        10   7   8  10       0             1     1  
3    ...    1    1       5         2  15  14  15       0             0     1  
4    ...    1    2       5         4   6  10  10       0             0     1  
..   ...  ...  ...     ...       ...  ..  ..  ..     ...           ...   ...  
390  ...    4    5       4        11   9   9   9       0             1     1  
391  ...    3    4       2         3  14  16  16       1             0     1  
392  ...    3    3       3         3  10   8   7       0             0     1  
393  ...    3    4       5         0  11  12  10       0             0     1  
394  ...    3    3       5         5   8   9   9       0             0     1  

[395 rows x 36 columns]

In [ ]:
# Path to the CSV file
csv_file_path = 'data.csv'

# Save DataFrame to CSV
df.to_csv(csv_file_path, index=False)

In [ ]:
df = df[['grade_A','high_absenses','count']]
df.head()

grade_A  high_absenses  count
0        0              0      1
1        0              0      1
2        0              1      1
3        0              0      1
4        0              0      1

In [ ]:
c = len(df[(df['high_absenses']==1) &(df['grade_A']==1)])
print(c)

5


In [ ]:
pd.pivot_table(
    df,
    values='count',
    index=['grade_A'],
    columns=['high_absenses'],
    aggfunc=np.size,
    fill_value=0
)

high_absenses    0   1
grade_A               
0              277  78
1               35   5

In our case:
P(A) is the probability of a grade of 80% or greater.
P(B) is the probability of missing 10 or more classes.
P(A|B) is the probability of a 80%+ grade, given missing 10 or more classes.

Calculations of parts:
P(A) = (35 + 5) / (35 + 5 + 277 + 78) = 0.10126582278481013
P(B) = (78 + 5) / (35 + 5 + 277 + 78) = 0.21012658227848102
P(A ∩ B) = 5 / (35 + 5 + 277 + 78) = 0.012658227848101266

P(A|B) = 0.012658227848101266/ 0.21012658227848102= 0.06

There we have it. The probability of getting at least an 80% final grade, given missing 10 or more classes is 6%.

Ques 2. **THE BIRTHDAY PROBLEM**
Compute the probability of getting a minimum of one overlapping birthday in a random group of 23 peoples.
Because ontaining random sample again and again is tedious task, we can do simulations on a computer with assumptions.
The birthdays are independent of each other. Each possible birthday has the same probability. There are only 365 possible birthdays (not 366, as ignoring the leap year.
Hint : in other words, we're modelling the process as drawing 23 independent samples from a discrete uniform distribution with parameter n = 365.

In [ ]:
from random import randint


NUM_PEOPLE = 23
NUM_POSSIBLE_BIRTHDAYS = 365
NUM_TRIALS = 10000


def generate_random_birthday():
    birthday = randint(1, NUM_POSSIBLE_BIRTHDAYS)
    return birthday


def generate_k_birthdays(k):
    birthdays = [generate_random_birthday() for _ in range(k)]
    return birthdays


def aloc(birthdays):
    unique_birthdays = set(birthdays)

    num_birthdays = len(birthdays)
    num_unique_birthdays = len(unique_birthdays)
    has_coincidence = (num_birthdays != num_unique_birthdays)

    return has_coincidence


def estimate_p_aloc():
    num_aloc = 0
    for _ in range(NUM_TRIALS):
        birthdays = generate_k_birthdays(NUM_PEOPLE)
        has_coincidence = aloc(birthdays)
        if has_coincidence:
            num_aloc += 1

    p_aloc = num_aloc / NUM_TRIALS
    return p_aloc


p_aloc = estimate_p_aloc()
print(f"Estimated P(ALOC) after {NUM_TRIALS} trials: {p_aloc}")

Estimated P(ALOC) after 10000 trials: 0.5051


**Ques 3. The weight of certain species of frog is uniformly distributed from 15 and 25 grams. if you randomly select a frog, what is the probability that the frog weights between 17 and 19 grams.**

In [ ]:
from scipy.stats import uniform
Pro = uniform.cdf(x=19, loc = 15, scale =10) -uniform.cdf(x=17, loc =15, scale=10)

In [ ]:
print(f"The probability that the frog weights between 17 and 19 grams is {Pro}")

The probability that the frog weights between 17 and 19 grams is 0.2


**Ques 4. Telecommunication Industry**
According to the Telecommunication Industry the average monthly cell phone bill is Rs. 1000 with a standard deviation of Rs. 200.

What is the probability that a randomly selected cell phone bill is more than Rs 1200? What is the probability that a randomly selected cell phone bill is between Rs 750 and Rs 1200? What is the probability that a randomly selected cell phone bill is no more than Rs 650? What is the amount above which lies top 15% of cell phone bills? What is the amount below which lies bottom 25% of cell phone bills?

Note: This is a problem of normal probability distribution. Though the distribution is not mentioned, in absence of any other information we assume normality in the population.

In [ ]:
from scipy import stats
a = stats.norm.cdf(1200,1000,200)
g = 1-a
print(g)

b= stats.norm.cdf(750,1000,200)
print(a-g)
stats.norm.ppf(.15,1000,200)

stats.norm.ppf(.25,1000,200)


0.15865525393145707
0.6826894921370859


865.1020499607837

Q5. Fruit problem
Suppose we own a fruit shop and on an average 3 customers arrive in the shop every 10 minutes. The mean rate here is 3 or λ = 3. Poisson probability distributions can help us answer questions like what is the probability that 5 customers will arrive in the next 10 mins?

In [ ]:
from scipy.stats import poisson
Value = poisson.cdf(k=5, mu=3)

In [ ]:
print(f"The probability that 5 customers will arrive in the next 10 mins is {Value}")

The probability that 5 customers will arrive in the next 10 mins is 0.9160820579686966
